pipelined/bevy_pbr2/src/render/pbr.wgsl

```diff
@@ -510,27 +533,31 @@ fn fragment(in: FragmentInput) -> [[location(0)]] vec4<f32> {
 
         var N: vec3<f32> = normalize(in.world_normal);
 
-        // FIXME: Normal maps need an additional vertex attribute and vertex stage output/fragment stage input
-        //        Just use a separate shader for lit with normal maps?
-        // #    ifdef STANDARDMATERIAL_NORMAL_MAP
-        //     vec3 T = normalize(v_WorldTangent.xyz);
-        //     vec3 B = cross(N, T) * v_WorldTangent.w;
-        // #    endif
+#ifdef VERTEX_TANGENTS
+#ifdef STANDARDMATERIAL_NORMAL_MAP
+        var T: vec3<f32> = normalize(in.world_tangent.xyz);
```

Comment:
Have you compared the cost of doing this here instead of in the vertex shader? This article points out an approach that seems like it could be much faster (because it removes matrix multiplication per fragment):
https://learnopengl.com/Advanced-Lighting/Normal-Mapping

Reply:
I haven't and my focus has been on correctness to start with, and if any compromises are going to be made for performance for example then I would definitely want to understand and document them.

To respond to this I have been trying to understand if there is some inaccuracy or error in the way the learnopengl.com tutorial does it as I have seen and experienced things suggesting that while informative, it is not the best reference. I prefer Foundations of Game Engine Development 2: Rendering, and/or Real-Time Rendering 4th Edition as reference material. The former pays attention to and explains small inaccuracies and how to fix them and seems to be a particularly good reference.

So a couple of comments:
* I feel like for now at least, transforming the normal map per-fragment tangent space normal to world space in the fragment shader and doing all lighting calculations in world space as before is intuitive to work with. I don't think transforming the lighting vectors (view vector, light vector) to tangent space in the vertex shader and doing calculations in tangent space is intuitive to work with, even if it may improve performance. We can try it to see if it makes a difference but the stress test would be a complex worlds with a lot of normal maps. Maybe the NVIDIA Orca Emerald Square City scene is a good candidate.
* I am wary of constructing the TBN matrix in the vertex shader either not being interpolated (I wonder how a mat3 should be interpolated sensibly? Maybe it just works, I don't know) or the interpolation introducing issues and then having to do calculations in the fragment shader to fix it up.

Foundations of Game Engine Development 2: Rendering, which I do trust as reference material that really pays attention to these kinds of details, does the following in the fragment shader:
```
uniform Texture2D normalMap;

float3 FetchNormalVector(float2 texcoord) {
  float2 m = texture(normalMap, texcoord).xy; 
  return (float3(m, sqrt(1.0 − m.x * m.x − m.y * m.y))); 
}

float3 FetchObjectNormalVector(float2 texcoord, float3 normal, float3 tangent, float sigma) {
  float3 m = FetchNormalVector(texcoord); 
  float3 n = normalize(normal); 
  float3 t = normalize(tangent − n * dot(tangent, n)); 
  float3 b = cross(normal, tangent) * sigma; 
  return (t * m.x + b * m.y + n * m.z); 
}
```
and notes:
> After interpolation, the normal and tangent vectors may not have unit length, and it’s possible that they are no longer perpendicular. To correct for this, we have to orthonormalize them in the pixel shader before we calculate the bitangent vector

So FGED2 recommends doing it basically the way I am doing it, and then additionally apply Gram-Schmidt to the world space tangent vector. I'll add that.

I think we could benchmark versus the calculate TBN in vertex shader and use it to transform the view and light vectors (and any others) there, etc in a separate step as I suspect it will have knock-on effects for other algorithms that work in world/view space. Is that ok?

Reply:
I've also just checked the left side of https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/NormalTangentMirrorTest and it looks correct to me, by the way.

![image](https://user-images.githubusercontent.com/302146/140312351-ad2a4255-0671-4826-8262-02a6693168d9.png)

Reply:
I have continued following rabbit holes to understand what to do here and it looks like UE4 interpolates the world space normal and tangent, calculates the bitangent simply using the cross product (no normalisation yet), samples the tangent space normal from the normal map, on mobile it doesn't normalise that for performance, elsewhere it does. The TBN matrix is constructed in the fragment shader, the tangent space normal is transformed to world space, and the result is normalised.

Now, I thought it interesting that they don't use Gram-Schmidt or try to ensure orthonormality across the interpolation. When constructing the TBN matrix, there is this comment:
```
    // Will not be orthonormal after interpolation. This perfectly matches xNormal.
    // Any mismatch with xNormal will cause distortions for baked normal maps.
```
It looks like UE4 uses https://github.com/mmikk/MikkTSpace to calculate normals, and that is used in Blender and xNormal. There is a discussion of what they're all doing here: http://www.mikktspace.com/ tl;dr calculating tangents and normal maps is practically an encoding process and the decoding process needs to be precisely the inverse of the encoding process in order to obtain the desired results. But different software does different things, so [Mikkelsen did a Masters thesis on it](http://image.diku.dk/projects/media/morten.mikkelsen.08.pdf) to try to design an approach that would be robust to things like the order of vertices supplied to process to generate tangents and normal maps, as well as the decoding process. This seems to be a possibly good pragmatic and practical solution that is worth investigating. Though perhaps it should be done separate to this PR?